plugin/pkg/admission/initialresources/hawkular.go

```diff
@@ -17,18 +17,195 @@ limitations under the License.
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	kubeClientCmd "k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
 )
 
-type hawkularSource struct{}
+type hawkularSource struct {
+	client       *metrics.Client
+	uri          *url.URL
+	useNamespace bool
+	modifiers    []metrics.Modifier
+}
+
+const (
+	containerImageTag string = "container_base_image"
+	descriptorTag     string = "descriptor_name"
+	separator         string = "/"
+
+	defaultServiceAccountFile = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+)
 
-func newHawkularSource() (dataSource, error) {
-	return nil, fmt.Errorf("hawkular source not implemented")
+// Get equivalent MetricDescriptor.Name used in the Heapster
+func heapsterEqName(kind api.ResourceName) string {
+	switch kind {
+	case api.ResourceCPU:
+		return "cpu/usage"
+	case api.ResourceMemory:
+		return "memory/usage"
+	default:
+		return ""
+	}
 }
 
-func (s *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
-	return 0, 0, fmt.Errorf("gcm source not implemented")
+// Create tagFilter query
+func tagQuery(kind api.ResourceName, image string, exactMatch bool) map[string]string {
+	q := make(map[string]string)
+
+	// Add here the descriptor_tag..
+	q[descriptorTag] = heapsterEqName(kind)
+
+	if exactMatch {
+		q[containerImageTag] = image
+	} else {
+		q[containerImageTag] = fmt.Sprintf("%s:*", strings.Split(image, ":")[0])
+	}
+
+	return q
+}
+
+// dataSource API
+
+func (self *hawkularSource) GetUsagePercentile(kind api.ResourceName, perc int64, image, namespace string, exactMatch bool, start, end time.Time) (int64, int64, error) {
+	q := tagQuery(kind, image, exactMatch)
+
+	m := make([]metrics.Modifier, len(self.modifiers), 2+len(self.modifiers))
+	copy(m, self.modifiers)
+
+	if namespace != "" {
+		m = append(m, metrics.Tenant(namespace))
+	}
+
+	p, err := metrics.ConvertToFloat64(perc)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	m = append(m, metrics.Filters(metrics.TagsFilter(q), metrics.BucketsFilter(1), metrics.StartTimeFilter(start), metrics.EndTimeFilter(end), metrics.PercentilesFilter([]float64{p})))
+
+	bp, err := self.client.ReadBuckets(metrics.Counter, m...)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return int64(bp[0].Percentiles[0].Value), int64(bp[0].Samples), nil
+}
+
+// Create new Hawkular Source. The uri follows the scheme from Heapster
+func newHawkularSource(uri string) (dataSource, error) {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &hawkularSource{
+		uri: u,
+	}
+	if err = d.init(); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// Almost equal to the Heapster initialization, try to keep them in the sync
+func (self *hawkularSource) init() error {
+	self.modifiers = make([]metrics.Modifier, 0)
+	p := metrics.Parameters{
+		Tenant: "heapster", // This data is stored by the heapster - for no-namespace hits
```

Comment:
No, this is only the default value (which you can override using a `tenant` parameter to hawkular), which is just the same for heapster and for this. Meaning that if user forgets to set it on Heapster and here, things would still work. If you change them on both, you can use heapster namespace for other things.